tests/helpers/debugadapter.py

```diff
@@ -9,40 +10,96 @@ class DebugAdapter(Closeable):
 
     PORT = 8888
 
+    # generic factories
+
     @classmethod
     def start(cls, argv, **kwargs):
-        def new_proc(argv, host, port):
+        def new_proc(argv, addr):
             argv = list(argv)
-            cls._ensure_addr(argv, host, port)
+            cls._ensure_addr(argv, addr)
             return Proc.start_python_module('ptvsd', argv)
         return cls._start(new_proc, argv, **kwargs)
 
     @classmethod
     def start_wrapper_script(cls, filename, argv, **kwargs):
-        def new_proc(argv, host, port):
+        def new_proc(argv, addr):
             return Proc.start_python_script(filename, argv)
         return cls._start(new_proc, argv, **kwargs)
 
+    # specific factory cases
+
+    @classmethod
+    def start_nodebug(cls, addr, name, kind='script'):
+        if kind == 'script':
+            argv = ['--nodebug', name]
+        elif kind == 'module':
+            argv = ['--nodebug', '-m', name]
+        else:
+            raise NotImplementedError
+        return cls.start(argv, addr=addr)
+
+    @classmethod
+    def start_as_server(cls, addr, *args, **kwargs):
+        addr = Address.as_server(*addr)
+        return cls._start_as(addr, *args, server=False, **kwargs)
+
+    @classmethod
+    def start_as_client(cls, addr, *args, **kwargs):
+        addr = Address.as_client(*addr)
+        return cls._start_as(addr, *args, server=False, **kwargs)
+
+    @classmethod
+    def start_for_attach(cls, addr, *args, **kwargs):
+        addr = Address.as_server(*addr)
+        return cls._start_as(addr, *args, server=True, **kwargs)
+
+    @classmethod
+    def _start_as(cls, addr, name, kind='script', extra=None, server=False):
+        argv = []
+        if server:
+            argv += ['--server']
+        if kind == 'script':
+            argv += [name]
+        elif kind == 'module':
+            argv += ['-m', name]
+        else:
+            raise NotImplementedError
+        if extra:
+            argv += list(extra)
+        return cls.start(argv, addr=addr)
+
+    @classmethod
+    def start_embedded(cls, addr, filename, redirect_output=True):
+        addr = Address.as_server(*addr)
+        with open(filename, 'r+') as scriptfile:
+            content = scriptfile.read()
+            if 'ptvsd.enable_attach' not in content:
+                raise NotImplementedError
```

Comment:
This looks more like an assert to me (the script *has* to invoke enable_attach, and this helps catch that instead of failing in a weird way). And if we intend to surface it for diagnostics, then perhaps treat it as invalid argument - so `ValueError` with a message?

Reply:
Conceivably it could be handled by automatically fixing the script.  That's why I left it NotImplementedError.  However, an assert (with a comment) is sufficient until it matters.